### PR TITLE
add pusher tests for all deep storages

### PR DESCRIPTION
**Pull request by fjy**
This PR depends on https://github.com/druid-io/druid/pull/2191

### CODE REVIEW

**Comment by fjy** [comment]
Fixes #2195

File: extensions/azure-extensions/pom.xml [hunk 1]
```diff
@@ -49,10 +50,22 @@
...
+            <version>${project.parent.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
             <groupId>org.easymock</groupId>
             <artifactId>easymock</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.beust</groupId>
```
Comment (fjy) [inline extensions/azure-extensions/pom.xml]: weird, could've sworn this was there before

File: extensions/azure-extensions/src/test/java/io/druid/storage/azure/AzureDataSegmentPusherTest.java [hunk 2]
```diff
@@ -93,7 +133,7 @@ public void uploadDataSegmentTest() throws StorageException, IOException, URISyn
     final File descriptorFile = new File("descriptor.json");
     final Map<String, String> azurePaths = pusher.getAzurePaths(dataSegment);
 
-    expect(azureAccountConfig.getContainer()).andReturn(containerName).times(3);
+    //expect(azureAccountConfig.getContainer()).andReturn(containerName).times(3);
```
Comment (fjy) [inline extensions/azure-extensions/src/test/java/io/druid/storage/azure/AzureDataSegmentPusherTest.java]: need to delete this

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureAccountConfig.java [hunk 3]
```diff
@@ -27,23 +27,23 @@
 public class AzureAccountConfig
 {
   @JsonProperty
-  private String protocol = "https";
+  public String protocol = "https";
```
Comment (drcrallen) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureAccountConfig.java]: Is it possible to leave these private and override the getters if needed in unit tests?

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java [hunk 4]
```diff
@@ -137,7 +129,9 @@ public DataSegment push(final File indexFilesDir, final DataSegment segment) thr
     log.info("Uploading [%s] to Azure.", indexFilesDir);
 
     final int version = SegmentUtils.getVersionFromDir(indexFilesDir);
-    final File compressedSegmentData = createCompressedSegmentDataFile(indexFilesDir);
+    final File zipOutFile = File.createTempFile("index", ".zip");
+    final long size = CompressionUtils.zip(indexFilesDir, zipOutFile);
```
Comment (drcrallen) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: Needs to be inside try to make sure cleanup happens

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureAccountConfig.java (hunk 3 above)
Comment (fjy) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureAccountConfig.java]: they are actually public everywhere and throughout the Druid codebase for configurations such as this one

File: extensions/azure-extensions/src/test/java/io/druid/storage/azure/AzureDataSegmentPusherTest.java (hunk 2 above)
Comment (drcrallen) [inline extensions/azure-extensions/src/test/java/io/druid/storage/azure/AzureDataSegmentPusherTest.java]: why was this removed?

File: extensions/cloudfiles-extensions/src/main/java/io/druid/storage/cloudfiles/CloudFilesDataSegmentPusherConfig.java [hunk 5]
```diff
@@ -31,18 +31,18 @@
 
   @JsonProperty
   @NotNull
-  private String region;
+  public String region;
```
Comment (drcrallen) [inline extensions/cloudfiles-extensions/src/main/java/io/druid/storage/cloudfiles/CloudFilesDataSegmentPusherConfig.java]: same question about overriding getters instead of making these public

File: extensions/cloudfiles-extensions/src/test/java/io/druid/storage/cloudfiles/CloudFilesDataSegmentPusherTest.java [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        Lists.<String>newArrayList(),
+        Lists.<String>newArrayList(),
+        new NoneShardSpec(),
+        0,
+        size
+    );
+
+    DataSegment segment = pusher.push(tmpDir, segmentToPush);
+
+    // Uncomment after https://github.com/druid-io/druid/pull/2191 is merged
```
Comment (drcrallen) [inline extensions/cloudfiles-extensions/src/test/java/io/druid/storage/cloudfiles/CloudFilesDataSegmentPusherTest.java]: should note dependency in PR master comment

**Comment by drcrallen** [comment]
For the temporary files and directories can you just use a TemporaryFolder `@Rule`?

File: extensions/azure-extensions/pom.xml (hunk 1 above)
Comment (drcrallen) [inline extensions/azure-extensions/pom.xml]: What is this being brought in for?

Comment (fjy) [inline extensions/azure-extensions/pom.xml]: not needed, will remove

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java (hunk 4 above)
Comment (drcrallen) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureDataSegmentPusher.java]: nvmnd, cleanup happens in uploadDataSegment

File: extensions/azure-extensions/src/test/java/io/druid/storage/azure/AzureDataSegmentPusherTest.java [hunk 7]
```diff
@@ -64,9 +70,43 @@
...
+    jsonMapper = new DefaultObjectMapper();
+  }
+
+  @Test
+  public void testPush() throws Exception
+  {
+    AzureDataSegmentPusher pusher = new AzureDataSegmentPusher(azureStorage, azureAccountConfig, jsonMapper);
+
+    // Create a mock segment on disk
+    File tmpDir = Files.createTempDir();
```
Comment (drcrallen) [inline extensions/azure-extensions/src/test/java/io/druid/storage/azure/AzureDataSegmentPusherTest.java]: Can this move to a TemporaryFolder `@Rule`?

File: extensions/cloudfiles-extensions/src/test/java/io/druid/storage/cloudfiles/CloudFilesDataSegmentPusherTest.java [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+    CloudFilesDataSegmentPusherConfig config = new CloudFilesDataSegmentPusherConfig();
+    config.region = "region";
+    config.container = "container";
+    config.basePath = "basePath";
+
+    CloudFilesDataSegmentPusher pusher = new CloudFilesDataSegmentPusher(api, config, new DefaultObjectMapper());
+
+    // Create a mock segment on disk
+    File tmpDir = Files.createTempDir();
```
Comment (drcrallen) [inline extensions/cloudfiles-extensions/src/test/java/io/druid/storage/cloudfiles/CloudFilesDataSegmentPusherTest.java]: create via TemporaryFolder

File: extensions/hdfs-storage/src/test/java/io/druid/storage/hdfs/HdfsDataSegmentPusherTest.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+ */
+public class HdfsDataSegmentPusherTest
+{
+  @Test
+  public void testPush() throws Exception
+  {
+    Configuration conf = new Configuration(true);
+
+    // Create a mock segment on disk
+    File tmpDir = Files.createTempDir();
```
Comment (drcrallen) [inline extensions/hdfs-storage/src/test/java/io/druid/storage/hdfs/HdfsDataSegmentPusherTest.java]: Create via TemporaryFolder

File: extensions/s3-extensions/src/test/java/io/druid/storage/s3/S3DataSegmentPusherTest.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+
+
+    S3DataSegmentPusherConfig config = new S3DataSegmentPusherConfig();
+    config.bucket = "bucket";
+    config.baseKey = "key";
+
+    S3DataSegmentPusher pusher = new S3DataSegmentPusher(s3Client, config, new DefaultObjectMapper());
+
+    // Create a mock segment on disk
+    File tmpDir = Files.createTempDir();
```
Comment (drcrallen) [inline extensions/s3-extensions/src/test/java/io/druid/storage/s3/S3DataSegmentPusherTest.java]: create via TemporaryFolder

File: extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureAccountConfig.java (hunk 3 above)
Comment (drcrallen) [inline extensions/azure-extensions/src/main/java/io/druid/storage/azure/AzureAccountConfig.java]: ok

**Comment by drcrallen** [comment]
Cool. :+1: after Travis

File: extensions/hdfs-storage/src/test/java/io/druid/storage/hdfs/HdfsDataSegmentPusherTest.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        0,
+        size
+    );
+
+    DataSegment segment = pusher.push(tempFolder.getRoot(), segmentToPush);
+
+    Assert.assertEquals(segmentToPush.getSize(), segment.getSize());
+
+    // cleanup
+    final File dir = new File(config.getStorageDirectory());
```
Comment (himanshug) [inline extensions/hdfs-storage/src/test/java/io/druid/storage/hdfs/HdfsDataSegmentPusherTest.java]: this should be taken care of TemporaryFolder itself

**Comment by himanshug** [comment]
:+1: after https://github.com/druid-io/druid/pull/2206#discussion_r48928237 is resolved and travis

File: extensions/azure-extensions/src/test/java/io/druid/storage/azure/AzureDataSegmentPusherTest.java [hunk 12]
```diff
@@ -21,27 +21,36 @@
...
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.io.Files;
 import com.metamx.common.MapUtils;
 import com.microsoft.azure.storage.StorageException;
+import io.druid.jackson.DefaultObjectMapper;
 import io.druid.segment.loading.DataSegmentPusherUtil;
 import io.druid.timeline.DataSegment;
 import io.druid.timeline.partition.NoneShardSpec;
+import junit.framework.Assert;
```
Comment (himanshug) [inline extensions/azure-extensions/src/test/java/io/druid/storage/azure/AzureDataSegmentPusherTest.java]: pls use org.junit.Assert instead of deprecated junit.framework.Assert
